README
NAME

    MooseX::AttributeShortcuts - Shorthand for common attribute options

VERSION

    This document describes version 0.033 of MooseX::AttributeShortcuts -
    released July 24, 2017 as part of MooseX-AttributeShortcuts.

SYNOPSIS

        package Some::Class;
    
        use Moose;
        use MooseX::AttributeShortcuts;
    
        # same as:
        #   is => 'ro', lazy => 1, builder => '_build_foo'
        has foo => (is => 'lazy');
    
        # same as: is => 'ro', writer => '_set_foo'
        has foo => (is => 'rwp');
    
        # same as: is => 'ro', builder => '_build_bar'
        has bar => (is => 'ro', builder => 1);
    
        # same as: is => 'ro', clearer => 'clear_bar'
        has bar => (is => 'ro', clearer => 1);
    
        # same as: is => 'ro', predicate => 'has_bar'
        has bar => (is => 'ro', predicate => 1);
    
        # works as you'd expect for "private": predicate => '_has_bar'
        has _bar => (is => 'ro', predicate => 1);
    
        # extending? Use the "Shortcuts" trait alias
        extends 'Some::OtherClass';
        has '+bar' => (traits => [Shortcuts], builder => 1, ...);
    
        # or...
        package Some::Other::Class;

DESCRIPTION

    Ever find yourself repeatedly specifying writers and builders, because
    there's no good shortcut to specifying them? Sometimes you want an
    attribute to have a read-only public interface, but a private writer.
    And wouldn't it be easier to just say builder => 1 and have the
    attribute construct the canonical _build_$name builder name for you?

    This package causes an attribute trait to be applied to all attributes
    defined to the using class. This trait extends the attribute option
    processing to handle the above variations. All attribute options as
    described in Moose or Class::MOP::Attribute remain usable, just as when
    this trait is not applied.

 Some Notes On History

    Moose has long had a lazy_build attribute option. It was once
    considered a best practice, but that has, ah, changed. This trait began
    as a desire to still leverage bits of lazy_build (and a tacit
    acknowledgment that fat-finger bugs rank among the most embarrassing,
    right up there with "the TV was unplugged the entire time").

    This author does not recommend you use lazy_build, unless you know
    exactly what you're doing (probably) and that it's a good idea
    (probably not).

    Nonetheless, this lazy_build option is why we set certain options the
    way we do below; while lazy_build in its entirety is not optimal, it
    had the right idea: regular, predictable accessor names for regular,
    predictable attribute options.

    As an example, just looking at the below it doesn't seem logical that:

        has _foo => (is => 'ro', clearer => 1);

    ...becomes:

        has _foo => (is => 'ro', clearer => '_clear_foo');

    After reading the lazy_build attribute option, however, we see that the
    choice had already been made for us.

USAGE

    This package automatically applies an attribute metaclass trait. Simply
    using this package causes the trait to be applied by default to your
    attribute's metaclasses.

EXTENDING A CLASS

    If you're extending a class and trying to extend its attributes as
    well, you'll find out that the trait is only applied to attributes
    defined locally in the class. This package exports a trait shortcut
    function Shortcuts that will help you apply this to the extended
    attribute:

        has '+something' => (traits => [Shortcuts], ...);

NEW ATTRIBUTE OPTIONS

    Unless specified here, all options defined by Moose::Meta::Attribute
    and Class::MOP::Attribute remain unchanged.

    Want to see additional options? Ask, or better yet, fork on GitHub and
    send a pull request. If the shortcuts you're asking for already exist
    in Moo or Mouse or elsewhere, please note that as it will carry
    significant weight.

    For the following, $name should be read as the attribute name; and the
    various prefixes should be read using the defaults.

 is => 'rwp'

    Specifying is => 'rwp' will cause the following options to be set:

        is     => 'ro'
        writer => "_set_$name"

    rwp can be read as "read + write private".

 is => 'lazy'

    Specifying is => 'lazy' will cause the following options to be set:

        is       => 'ro'
        builder  => "_build_$name"
        lazy     => 1

    NOTE: Since 0.009 we no longer set init_arg => undef if no init_arg is
    explicitly provided. This is a change made in parallel with Moo, based
    on a large number of people surprised that lazy also made one's
    init_def undefined.

 is => 'lazy', default => ...

    Specifying is => 'lazy' and a default will cause the following options
    to be set:

        is       => 'ro'
        lazy     => 1
        default  => ... # as provided

    That is, if you specify is => 'lazy' and also provide a default, then
    we won't try to set a builder, as well.

 builder => 1

    Specifying builder => 1 will cause the following options to be set:

        builder => "_build_$name"

 builder => sub { ... }

    Passing a coderef to builder will cause that coderef to be installed in
    the class this attribute is associated with the name you'd expect, and
    builder => 1 to be set.

    e.g., in your class (or role),

        has foo => (is => 'ro', builder => sub { 'bar!' });

    ...is effectively the same as...

        has foo => (is => 'ro', builder => '_build_foo');
        sub _build_foo { 'bar!' }

    The behaviour of this option in roles changed in 0.030, and the builder
    methods will be installed in the role itself. This means you can
    alias/exclude/etc builder methods in roles, just as you can with any
    other method.

 clearer => 1

    Specifying clearer => 1 will cause the following options to be set:

        clearer => "clear_$name"

    or, if your attribute name begins with an underscore:

        clearer => "_clear$name"

    (that is, an attribute named _foo would get _clear_foo)

 predicate => 1

    Specifying predicate => 1 will cause the following options to be set:

        predicate => "has_$name"

    or, if your attribute name begins with an underscore:

        predicate => "_has$name"

    (that is, an attribute named _foo would get _has_foo)

 trigger => 1

    Specifying trigger => 1 will cause the attribute to be created with a
    trigger that calls a named method in the class with the options passed
    to the trigger. By default, the method name the trigger calls is the
    name of the attribute prefixed with _trigger_.

    e.g., for an attribute named foo this would be equivalent to:

        trigger => sub { shift->_trigger_foo(@_) }

    For an attribute named _foo:

        trigger => sub { shift->_trigger__foo(@_) }

    This naming scheme, in which the trigger is always private, is the same
    as the builder naming scheme (just with a different prefix).

 handles => { foo => sub { ... }, ... }

    Creating a delegation with a coderef will now create a new, "custom
    accessor" for the attribute. These coderefs will be installed and
    called as methods on the associated class (just as readers, writers,
    and other accessors are), and will have the attribute metaclass
    available in $_. Anything the accessor is called with it will have
    access to in @_, just as you'd expect of a method.

    e.g., the following example creates an attribute named bar with a
    standard reader accessor named bar and two custom accessors named foo
    and foo_too.

        has bar => (
    
            is      => 'ro',
            isa     => 'Int',
            handles => {
    
                foo => sub {
                    my $self = shift @_;
    
                    return $_->get_value($self) + 1;
                },
    
                foo_too => sub {
                    my $self = shift @_;
    
                    return $self->bar + 1;
                },
    
                # ...as you'd expect.
                bar => 'bar',
            },
        );

    ...and later,

    Note that in this example both foo() and foo_too() do effectively the
    same thing: return the attribute's current value plus 1. However, foo()
    accesses the attribute value directly through the metaclass, the pros
    and cons of which this author leaves as an exercise for the reader to
    determine.

    You may choose to use the installed accessors to get at the attribute's
    value, or use the direct metaclass access, your choice.

ANONYMOUS SUBTYPING AND COERCION

        "Abusus non tollit usum."

    Note that we create new, anonymous subtypes whenever the constraint or
    coercion options are specified in such a way that the Shortcuts trait
    (this one) is invoked. It's fully supported to use both constraint and
    coerce options at the same time.

    This facility is intended to assist with the creation of one-off type
    constraints and coercions. It is not possible to deliberately reuse the
    subtypes we create, and if you find yourself using a particular isa /
    constraint / coerce option triplet in more than one place you should
    really think about creating a type that you can reuse. MooseX::Types
    provides the facilities to easily do this, or even a simple constant
    definition at the package level with an anonymous type stashed away for
    local use.

 isa => sub { ... }

        has foo => (
            is  => 'rw',
            # $_ == $_[0] == the value to be validated
            isa => sub { die unless $_[0] == 1 },
        );
    
        # passes constraint
        $thing->foo(1);
    
        # fails constraint
        $thing->foo(5);

    Given a coderef, create a type constraint for the attribute. This
    constraint will fail if the coderef dies, and pass otherwise.

    Astute users will note that this is the same way Moo constraints work;
    we use MooseX::Meta::TypeConstraint::Mooish to implement the
    constraint.

 isa_instance_of => ...

    Given a package name, this option will create an isa type constraint
    that requires the value of the attribute be an instance of the class
    (or a descendant class) given. That is,

        has foo => (is => 'ro', isa_instance_of => 'SomeThing');

    ...is effectively the same as:

        use Moose::TypeConstraints 'class_type';
        has foo => (
            is  => 'ro',
            isa => class_type('SomeThing'),
        );

    ...but a touch less awkward.

 isa => ..., constraint => sub { ... }

    Specifying the constraint option with a coderef will cause a new
    subtype constraint to be created, with the parent type being the type
    specified in the isa option and the constraint being the coderef
    supplied here.

    For example, only integers greater than 10 will pass this attribute's
    type constraint:

        # value must be an integer greater than 10 to pass the constraint
        has thinger => (
            isa        => 'Int',
            constraint => sub { $_ > 10 },
            # ...
        );

    Note that if you supply a constraint, you must also provide an isa.

 isa => ..., constraint => sub { ... }, coerce => 1

    Supplying a constraint and asking for coercion will "Just Work", that
    is, any coercions that the isa type has will still work.

    For example, let's say that you're using the File type constraint from
    MooseX::Types::Path::Class, and you want an additional constraint that
    the file must exist:

        has thinger => (
            is         => 'ro',
            isa        => File,
            constraint => sub { !! $_->stat },
            coerce     => 1,
        );

    thinger will correctly coerce the string "/etc/passwd" to a
    Path::Class:File, and will only accept the coerced result as a value if
    the file exists.

 coerce => [ Type => sub { ...coerce... }, ... ]

    Specifying the coerce option with a hashref will cause a new subtype to
    be created and used (just as with the constraint option, above), with
    the specified coercions added to the list. In the passed hashref, the
    keys are Moose types (well, strings resolvable to Moose types), and the
    values are coderefs that will coerce a given type to our type.

        has bar => (
            is     => 'ro',
            isa    => 'Str',
            coerce => [
                Int    => sub { "$_"                       },
                Object => sub { 'An instance of ' . ref $_ },
            ],
        );

INTERACTIONS WITH OTHER ATTRIBUTE TRAITS

    Sometimes attribute traits interact in surprising ways. This trait is
    well behaved; if you have discovered any interactions with other traits
    (good, bad, indifferent, etc), please report this
    <https://github.com/RsrchBoy/moosex-attributeshortcuts/issues/new> so
    that it can be worked around, fixed, or documented, as appropriate.

 MooseX::SemiAffordanceAccessor

    MooseX::SemiAffordanceAccessor changes how the is = 'rw'> and accessor
    = ...> attribute options work. If our trait detects that an attribute
    has had the MooseX::SemiAffordanceAccessor attribute trait applied,
    then we change our behaviour to conform to its expectations:

      * is => 'rwp'

      This is now effectively equivalent to:

          is => 'ro',
          writer => '_set_...',

      * writer_prefix is ignored


Changes
Revision history for MooseX-AttributeShortcuts

{{$NEXT}}
	* Fix a documentation error (thanks, @2shortplanks! issue #21)

0.037     2017-11-20 16:48:24 CST-0600
	* Add `init_arg` shortcuts.

0.036     2017-10-31 21:34:35 CDT-0500
	* When creating an inline builder method's metaclass, take a copy of the
	  attribute's own definition context; previously we were inadvertently
	  modifying the attribute's definition context with rather less than
	  helpful results.

0.035     2017-09-22 19:54:38 CDT-0500
	* Rebrand "anonymous builders" as "inline builders"
	* Flavor the method metaclass with a trait that provides for definition
	  context, and install inline builders using it.  This will make it easier
	  to use the method metaclasses to see where the builder methods came from
	  (if you're in to that sort of thing).

0.034     2017-07-25 19:18:32 CDT-0500
	* Make optional requirements optional.  Thanks @ppisar!  Closes #19

0.033     2017-07-24 21:52:45 CDT-0500
	* Detect if MooseX::SemiAffordanceAccessor has been applied to the
	  attribute, and conform our behaviour to it if necessary.  Add some docs,
	  tests around this.

0.032     2017-06-13 21:59:13 CDT-0500
	* Move builder method name generation into its own method, provided by the
	  common 'HasAnonBuilder' role, to allow for easier overriding when
	  organizations decide to use non-standard builder method names that
	  cannot be handled by changing the traits' "builder_prefix" option.

0.031     2017-05-30 22:32:27 CDT-0500
	* Fix tests to run without '.' in @INC; fixes #17.
	* Additional tests, documentation, and examples.

0.030-TRIAL  2017-05-08 22:58:52 CDT-0500
	* Move our attribute trait out into its own file.
	* Anonymous builder methods for attributes in roles are now created at the
	  time the attribute is defined in the role, not when the role (attribute)
	  is applied to a class.  That is, using 'builder => sub { ... }' in a
	  role will result in the role having the builder method.  Because of
	  this, we can now treat those methods like other role methods: aliasing,
	  excluding, etc.

0.029     2017-05-08 22:53:18 CDT-0500
	* Redo a number of tests, to improve coverage and deal with the IMPENDING
	  DOOM of . in @INC
	* Add a "history" note referencing lazy_build
	* Additional metaclass/trait application tests
	* Add a 'trigger_method' attribute, to stash away the expected name of the
	  method our generated trigger will call in the associated class
	* Use List::Util instead of List::AllUtils+List::MoreUtils (thanks,
	  ether! #15)
	* Clarify some points in the documentation (thanks, Olaf! #10)
	* Drop some older methods: mi(), strengthen/weaken_value()

0.028     2015-04-07 20:45:00 PDT-0700
	* 0.028.  *le sigh*

0.027_02  2015-04-07 20:43:17 PDT-0700
	* Moose 1.14 is now the minimum supported Moose.

0.027_01  2015-03-28 22:02:17 PDT-0700
	* Break out our $_process_options into more... succinct... methods.
	* Add Moo-style type constraints.

0.027     2015-03-10 19:27:06 PDT-0700
	* Stop using autobox, after relentless persecution by The Cabal ;)
	* Push out our warnings for those using undocumented bits, as we haven't
	  released a non-dev/trial version in *ages*

0.026-TRIAL  2014-12-02 19:52:59 PST-0800
	* Keep our comments dumb, even in tests. (thanks, @haarg!)

0.025-TRIAL  2014-05-22 12:51:25 PDT-0700
	* This is a TRIAL release, as I'm not entirely sure of the wisdom and
	  implementation of the following.
	* Allow for the creation of additional associated methods by hijacking the
	  delegation system, a la native traits: handles => { ... => sub { ... } }

0.024     2014-05-02 12:50:57 PDT-0700
	* Make the undocumented isa_* options "deprecated" ...which seems like
	  better karma than no warning, as people really shouldn't be using these
	  anyways, but better nice than to have it done to me someday :)

0.023     2014-04-04 21:23:33 PDT-0700
	* Add isa_instance_of attribute option.

0.022     2013-09-29 00:08:39 PDT-0700
	* No code changes -- released with refreshed test suite

0.021     2013-09-08 23:32:09 PDT-0700
	* Drop MXCPA entirely, and incorporate the subtyping functionality

0.020_01  2013-08-26 13:33:44 PDT-0700
	* Keep MXCPA quiet, for now.  This is evil (hence the dev release), but
	  should keep things status-quo until we properly integrate MXCPA
	  functionality into MXAS proper.

0.020     2013-08-19 21:51:02 PDT-0700
	* Bump CoercePerAttribute version requirement to 1.000 and update docs

0.019     2013-04-20 21:50:05 PDT-0700
	* Better support inline subtyping and coercions (people who are inclined
	  to shoot themselves in their foot will do it regardless)
	* Additional tests for coercions

0.018     2013-01-09 10:20:25 PST8PDT
	* ~~ TRIAL ~~
	* If a constraint coderef is given and coercion is requested, we now copy
	  the parent type's coercions to the anonymous child type.
	* Better document the constraint option and how it interacts with simple
	  coercion.

0.017     2012-10-28 20:53:52 PST8PDT
	* make actually make classes immutable during tests
	* add initial primitive anonymous inline subtyping support; this needs to
	  be evaluated with the outstanding pull request (issue #6) but I need
	  this Right Now

0.016     2012-09-08 21:41:38 America/Los_Angeles
	* provide for "negative shortcuts" for clearer and predicate, but don't
	  document for the moment

0.015     2012-08-26 11:48:29 America/Los_Angeles
	* add 'builder => sub { ... }' shortcut [gh #4]

0.014     2012-08-16 17:26:09 America/Los_Angeles
	* explictly test for definedness rather than truthiness when determining
	  if someone has also specified a default along with is => lazy

0.013     2012-07-15 22:11:37 America/Los_Angeles
	* Restore tests lost from (d11e67a) -- no functional / code changes!

0.012     2012-05-02 11:02:10 America/Los_Angeles
	* TRIAL/dev release
	* Don't try to guess what sort of metaclass we need; just die if we don't
	  have one
	* Misc cleanups

0.011     2012-04-30 13:41:05 America/Los_Angeles

lib/MooseX/AttributeShortcuts.pm
package MooseX::AttributeShortcuts;

# ABSTRACT: Shorthand for common attribute options

use strict;
use warnings;

use namespace::autoclean;

use Moose 1.14 ();
use Moose::Exporter;
use Moose::Meta::TypeConstraint;
use Moose::Util::MetaRole;
use Moose::Util::TypeConstraints;

use MooseX::AttributeShortcuts::Trait::Attribute;
use MooseX::AttributeShortcuts::Trait::Role::Attribute;

my ($import, $unimport, $init_meta) = Moose::Exporter->build_import_methods(
    install => [ 'unimport' ],
    trait_aliases => [
        [ 'MooseX::AttributeShortcuts::Trait::Attribute' => 'Shortcuts' ],
    ],
);

my $role_params;

sub import {
    my ($class, %args) = @_;

    $role_params = {};
    do { $role_params->{$_} = delete $args{"-$_"} if exists $args{"-$_"} }
        for qw{ writer_prefix builder_prefix prefixes };

    @_ = ($class, %args);
    goto &$import;
}

sub init_meta {
    my ($class_name, %args) = @_;
    my $params = delete $args{role_params} || $role_params || undef;
    undef $role_params;

    # Just in case we do ever start to get an $init_meta from ME
    $init_meta->($class_name, %args)
        if $init_meta;

    # make sure we have a metaclass instance kicking around
    my $for_class = $args{for_class};
    die "Class $for_class has no metaclass!"
        unless Class::MOP::class_of($for_class);

    # If we're given parameters to pass on to construct a role with, we build
    # it out here rather than pass them on and allowing apply_metaroles() to
    # handle it, as there are Very Loud Warnings about how parameterized roles
    # are non-cacheable when generated on the fly.

    ### $params
    my $trait
        = ($params && scalar keys %$params)
        ? MooseX::AttributeShortcuts::Trait::Attribute
            ->meta
            ->generate_role(parameters => $params)
        : 'MooseX::AttributeShortcuts::Trait::Attribute'
        ;

    my $role_attribute_trait
        = ($params && exists $params->{builder_prefix})
        ? MooseX::AttributeShortcuts::Trait::Role::Attribute
            ->meta
            ->generate_role(
                parameters => { builder_prefix => $params->{builder_prefix} },
            )
        : 'MooseX::AttributeShortcuts::Trait::Role::Attribute'
        ;

    Moose::Util::MetaRole::apply_metaroles(
        # TODO add attribute trait here to create builder method if found
        for                          => $for_class,
        class_metaroles              => { attribute         => [ $trait ] },
        role_metaroles               => {
            applied_attribute => [ $trait ],
            # attribute => [ 'MooseX::AttributeShortcuts::Trait::Role::Attribute' ],
            attribute         => [ $role_attribute_trait ],
        },
        parameter_metaroles          => { applied_attribute => [ $trait ] },
        parameterized_role_metaroles => { applied_attribute => [ $trait ] },
    );

    return Class::MOP::class_of($for_class);
}

1;

__END__

=for :stopwords GitHub attribute's isa one's rwp SUBTYPING foo

=for Pod::Coverage init_meta

=head1 SYNOPSIS

    package Some::Class;

    use Moose;
    use MooseX::AttributeShortcuts;

    # same as:
    #   is => 'ro', lazy => 1, builder => '_build_foo'
    has foo => (is => 'lazy');

    # same as: is => 'ro', writer => '_set_foo'
    has foo => (is => 'rwp');

    # same as: is => 'ro', builder => '_build_bar'
    has bar => (is => 'ro', builder => 1);

    # same as: is => 'ro', clearer => 'clear_bar'
    has bar => (is => 'ro', clearer => 1);

    # same as: is => 'ro', predicate => 'has_bar'
    has bar => (is => 'ro', predicate => 1);

    # works as you'd expect for "private": predicate => '_has_bar'
    has _bar => (is => 'ro', predicate => 1);

    # extending? Use the "Shortcuts" trait alias
    extends 'Some::OtherClass';
    has '+bar' => (traits => [Shortcuts], builder => 1, ...);

=head1 DESCRIPTION

Ever find yourself repeatedly specifying writers and builders, because there's
no good shortcut to specifying them?  Sometimes you want an attribute to have
a read-only public interface, but a private writer.  And wouldn't it be easier
to just say C<< builder => 1 >> and have the attribute construct the canonical
C<_build_$name> builder name for you?

This package causes an attribute trait to be applied to all attributes defined
to the using class.  This trait extends the attribute option processing to
handle the above variations.  All attribute options as described in L<Moose>
or L<Class::MOP::Attribute> remain usable, just as when this trait is not
applied.

=head2 Some Notes On History

Moose has long had a L<lazy_build attribute option|Moose/lazy_build>.  It was
once considered a best practice, but that has, ah, changed.  This trait began
as a desire to still leverage bits of C<lazy_build> (and a tacit
acknowledgment that fat-finger bugs rank among the most embarrassing, right up
there with "the TV was unplugged the entire time").

This author does not recommend you use C<lazy_build>, unless you know exactly
what you're doing (probably) and that it's a good idea (probably not).

Nonetheless, this C<lazy_build> option is why we set certain options the way
we do below; while C<lazy_build> in its entirety is not optimal, it had the
right idea: regular, predictable accessor names for regular, predictable
attribute options.

As an example, just looking at the below it doesn't seem logical that:

    has _foo => (is => 'ro', clearer => 1);

...becomes:

    has _foo => (is => 'ro', clearer => '_clear_foo');

After reading the L<lazy_build attribute option|Moose/lazy_build>,
however, we see that the choice had already been made for us.

=head1 USAGE

This package automatically applies an attribute metaclass trait.  Simply using
this package causes the trait to be applied by default to your attribute's
metaclasses.

=head1 EXTENDING A CLASS

If you're extending a class and trying to extend its attributes as well,
you'll find out that the trait is only applied to attributes defined locally
in the class.  This package exports a trait shortcut function C<Shortcuts>
that will help you apply this to the extended attribute:

    has '+something' => (traits => [Shortcuts], ...);

=head1 NEW ATTRIBUTE OPTIONS

Unless specified here, all options defined by L<Moose::Meta::Attribute> and
L<Class::MOP::Attribute> remain unchanged.

Want to see additional options?  Ask, or better yet, fork on GitHub and send
a pull request.  If the shortcuts you're asking for already exist in L<Moo> or
L<Mouse> or elsewhere, please note that as it will carry significant weight.

For the following, C<$name> should be read as the attribute name; and the
various prefixes should be read using the defaults.

=head2 is => 'rwp'

Specifying C<is =E<gt> 'rwp'> will cause the following options to be set:

    is     => 'ro'
    writer => "_set_$name"

rwp can be read as "read + write private".

=head2 is => 'lazy'

Specifying C<is =E<gt> 'lazy'> will cause the following options to be set:

    is       => 'ro'
    builder  => "_build_$name"
    lazy     => 1

B<NOTE:> Since 0.009 we no longer set C<init_arg =E<gt> undef> if no C<init_arg>
is explicitly provided.  This is a change made in parallel with L<Moo>, based
on a large number of people surprised that lazy also made one's C<init_def>
undefined.

=head2 is => 'lazy', default => ...

Specifying C<is =E<gt> 'lazy'> and a default will cause the following options to be
set:

    is       => 'ro'
    lazy     => 1
    default  => ... # as provided

That is, if you specify C<is =E<gt> 'lazy'> and also provide a C<default>, then
we won't try to set a builder, as well.

=head2 builder => 1

Specifying C<builder =E<gt> 1> will cause the following options to be set:

    builder => "_build_$name"

=head2 builder => sub { ... }

Passing a coderef to builder will cause that coderef to be installed in the
class this attribute is associated with the name you'd expect, and
C<builder =E<gt> 1> to be set.

e.g., in your class (or role),

    has foo => (is => 'ro', builder => sub { 'bar!' });

...is effectively the same as...

    has foo => (is => 'ro', builder => '_build_foo');
    sub _build_foo { 'bar!' }

The behaviour of this option in roles changed in 0.030, and the builder
methods will be installed in the role itself.  This means you can
alias/exclude/etc builder methods in roles, just as you can with any other
method.

=head2 clearer => 1

Specifying C<clearer =E<gt> 1> will cause the following options to be set:

    clearer => "clear_$name"

or, if your attribute name begins with an underscore:

    clearer => "_clear$name"

(that is, an attribute named C<_foo> would get C<_clear_foo>)

=head2 predicate => 1

Specifying C<predicate =E<gt> 1> will cause the following options to be set:

    predicate => "has_$name"

or, if your attribute name begins with an underscore:

    predicate => "_has$name"

(that is, an attribute named C<_foo> would get C<_has_foo>)

=head2 init_arg => 1 / -1

This is a somewhat esoteric shortcut; you probably don't want to use this (or
even read this section).

Specifying C<init_arg =E<gt> 1> will cause the following options to be set:

    # attribute: "name"
    init_arg => 'name'

    # or, attribute: "_name"
    init_arg => '_name'

...while C<init_arg =E<gt> -1> will cause the following options to be set:

    # attribute: "name"
    init_arg => '_name'

    # or, attribute: "_name"
    init_arg => 'name'

=head2 trigger => 1

Specifying C<trigger =E<gt> 1> will cause the attribute to be created with a trigger
that calls a named method in the class with the options passed to the trigger.
By default, the method name the trigger calls is the name of the attribute
prefixed with C<_trigger_>.

e.g., for an attribute named C<foo> this would be equivalent to:

    trigger => sub { shift->_trigger_foo(@_) }

For an attribute named C<_foo>:

    trigger => sub { shift->_trigger__foo(@_) }

This naming scheme, in which the trigger is always private, is the same as the
builder naming scheme (just with a different prefix).

=head2 handles => { foo => sub { ... }, ... }

Creating a delegation with a coderef will now create a new, "custom accessor"
for the attribute.  These coderefs will be installed and called as methods on
the associated class (just as readers, writers, and other accessors are), and
will have the attribute metaclass available in C<$_>.  Anything the accessor
is called with it will have access to in C<@_>, just as you'd expect of a
method.

e.g., the following example creates an attribute named C<bar> with a standard
reader accessor named C<bar> and two custom accessors named C<foo> and
C<foo_too>.

    has bar => (

        is      => 'ro',
        isa     => 'Int',
        handles => {

            foo => sub {
                my $self = shift @_;

                return $_->get_value($self) + 1;
            },

            foo_too => sub {
                my $self = shift @_;

                return $self->bar + 1;
            },

            # ...as you'd expect.
            bar => 'bar',
        },
    );

...and later,

Note that in this example both foo() and foo_too() do effectively the same
thing: return the attribute's current value plus 1.  However, foo() accesses
the attribute value directly through the metaclass, the pros and cons of
which this author leaves as an exercise for the reader to determine.

You may choose to use the installed accessors to get at the attribute's value,
or use the direct metaclass access, your choice.

=head1 ANONYMOUS SUBTYPING AND COERCION

    "Abusus non tollit usum."

Note that we create new, anonymous subtypes whenever the constraint or
coercion options are specified in such a way that the Shortcuts trait (this
one) is invoked.  It's fully supported to use both constraint and coerce
options at the same time.

This facility is intended to assist with the creation of one-off type
constraints and coercions.  It is not possible to deliberately reuse the
subtypes we create, and if you find yourself using a particular isa /
constraint / coerce option triplet in more than one place you should really
think about creating a type that you can reuse.  L<MooseX::Types> provides
the facilities to easily do this, or even a simple L<constant> definition at
the package level with an anonymous type stashed away for local use.

=head2 isa => sub { ... }

    has foo => (
        is  => 'rw',
        # $_ == $_[0] == the value to be validated
        isa => sub { die unless $_[0] == 1 },
    );

    # passes constraint
    $thing->foo(1);

    # fails constraint
    $thing->foo(5);

Given a coderef, create a type constraint for the attribute.  This constraint
will fail if the coderef dies, and pass otherwise.

Astute users will note that this is the same way L<Moo> constraints work; we
use L<MooseX::Meta::TypeConstraint::Mooish> to implement the constraint.

=head2 isa_instance_of => ...

Given a package name, this option will create an C<isa> type constraint that
requires the value of the attribute be an instance of the class (or a
descendant class) given.  That is,

    has foo => (is => 'ro', isa_instance_of => 'SomeThing');

...is effectively the same as:

    use Moose::Util::TypeConstraints 'class_type';
    has foo => (
        is  => 'ro',
        isa => class_type('SomeThing'),
    );

...but a touch less awkward.

=head2 isa => ..., constraint => sub { ... }

Specifying the constraint option with a coderef will cause a new subtype
constraint to be created, with the parent type being the type specified in the
C<isa> option and the constraint being the coderef supplied here.

For example, only integers greater than 10 will pass this attribute's type
constraint:

    # value must be an integer greater than 10 to pass the constraint
    has thinger => (
        isa        => 'Int',
        constraint => sub { $_ > 10 },
        # ...
    );

Note that if you supply a constraint, you must also provide an C<isa>.

=head2 isa => ..., constraint => sub { ... }, coerce => 1

Supplying a constraint and asking for coercion will "Just Work", that is, any
coercions that the C<isa> type has will still work.

For example, let's say that you're using the C<File> type constraint from
L<MooseX::Types::Path::Class>, and you want an additional constraint that the
file must exist:

    has thinger => (
        is         => 'ro',
        isa        => File,
        constraint => sub { !! $_->stat },
        coerce     => 1,
    );

C<thinger> will correctly coerce the string "/etc/passwd" to a
C<Path::Class:File>, and will only accept the coerced result as a value if
the file exists.

=head2 coerce => [ Type => sub { ...coerce... }, ... ]

Specifying the coerce option with a hashref will cause a new subtype to be
created and used (just as with the constraint option, above), with the
specified coercions added to the list.  In the passed hashref, the keys are
Moose types (well, strings resolvable to Moose types), and the values are
coderefs that will coerce a given type to our type.

    has bar => (
        is     => 'ro',
        isa    => 'Str',
        coerce => [
            Int    => sub { "$_"                       },
            Object => sub { 'An instance of ' . ref $_ },
        ],
    );

=head1 INTERACTIONS WITH OTHER ATTRIBUTE TRAITS

Sometimes attribute traits interact in surprising ways.  This trait is well
behaved; if you have discovered any interactions with other traits (good, bad,
indifferent, etc), please
L<report this|https://github.com/RsrchBoy/moosex-attributeshortcuts/issues/new>
so that it can be worked around, fixed, or documented, as appropriate.

=head2 MooseX::SemiAffordanceAccessor

L<MooseX::SemiAffordanceAccessor> changes how the C<< is => 'rw' >> and
C<< accessor => ... >> attribute options work.  If our trait detects that an
attribute has had the
L<MooseX::SemiAffordanceAccessor attribute trait|MooseX::SemiAffordanceAccessor::Role::Attribute>
applied, then we change our behaviour to conform to its expectations:

=begin :list

* C<< is => 'rwp' >>

This:

    has  foo => (is => 'rwp');
    has _bar => (is => 'rwp');

...is now effectively equivalent to:

    has foo  => (is => 'ro', writer => '_set_foo');
    has _bar => (is => 'ro', writer => '_set_bar')

* C<-writer_prefix> is ignored

...as MooseX::SemiAffordanceAccessor has its own specific ideas as to how
writers should look.

=end :list

=head1 SEE ALSO

Moo
MooseX::Types
MooseX::SemiAffordanceAccessor

=cut
